File: top-interview-questions-medium/Array and Strings/longest-substring-without-repeating-characters/3.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
  let len = s.length
  if (len < 2) return len

  let max = 0
  let last = 0
  for (let i = 1; i < len; i++) {
    let before = s.lastIndexOf(s[i], i - 1)
    if (before < last) {
      max = Math.max(i - last + 1, max)
    } else {
      max = Math.max(i - before, max)
      last = before + 1
    }
  }
  return max
}

File: top-interview-questions-medium/Array and Strings/longest-substring-without-repeating-characters/3.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s) < 2: return len(s)

        maxL, last = 0, 0
        for i in range(1, len(s)):
            before = s.rfind(s[i], 0, i)
            if before < last:
                maxL = max(i - last + 1, maxL)
            else:
                maxL = max(i - before, maxL)
                last = before + 1
        return maxL
